### ISSUE TITLE
Log shallow config resolution

### ISSUE THREAD

**Issue by rarkins** [renovatebot/renovate#15827]
### What would you like Renovate to be able to do?

Today Renovate logs a repo's config (e.g. `renovate.json`) as well as whether it needs migrating. But in many cases a repo may just be extending a central config which contains all the relevant configuration details. The problem is that when we view the debug logs, we are essentially "blind" to this central config.

There exists the `printConfig` option, but this prints the entire resolved config, including e.g. `config:base` and `group:monorepos`, which are really long. So the user's config can get lost in the noise if we log that.

Instead, I would like to also log a "shallow" preset resolution result which resolves only non-internal presets while leaving internal presets as-is (inside whatever `extends` array(s) they are in).

### If you have any ideas on how this should be implemented, please tell us here.

We could add a new flag to resolvePresets() which allows for internal-only resolution. Such resolution skips any internal presets and leaves them as-is.

Then once we have detected a repo's config we can then call this function with internal-only resolution, to produce the "shallow" config. This shallow config would then be logged.

The shallow config would not be used after it is logged. Important:
- Need to make sure the resolve presets function is immutable (we don't want resolving to alter the original config)
- Full (existing) resolution should be done afterwards based on the *original* repo config and no based on the shallow resolved config. This is because we need to make sure that we resolve in the correct order

### Is this a feature you are interested in implementing yourself?

No

**Comment by rarkins** [comment]
The new flag is internal only, to adjust the behavior of resolvePresets(). This shallow resolution and debugging should be default behavior and not need any new config option.

**Comment by hasanwhitesource** [comment]
Just to perfectly understand the expected behavior. I have these 2 repos for example : 
```
// github.com/example/repo-a

{
	"packageRules": [
		{
			"matchPackageNames": [
				"@floating-ui/react-dom"
			],
			"allowedVersions": "<0.6.3"
		},
		{
			"updateTypes": [
				"major"
			],
			"labels": [
				"UPDATE-MAJOR"
			]
		}
	]
}

```

```
// github.com/example/repo-b

{
	"extends": [
		"config:base",
		"github>example/repo-a:renovate.json"
	],
	"ignoreDeps": [
		"next"
	]
}

```
```
// github.com/example/repo-c (without `config:base`)
{
	"extends": [
		"github>example/repo-a:renovate.json"
	],
	"ignoreDeps": [
		"next"
	]
}
```
Using the flag `printConfig` we get an additional print `"msg": "Full resolved config including presets"`

- for `repo-b` it prints a total of ~4.1K lines of config.
- for `repo-c` (without `config:base`) it prints a total of ~700 lines of config.

Talking about the new **shallow** behavior print:

- Is the expected print supposed to be the combined config without the extend block ?
- The same result for both `repo-b` and `repo-c` ?

Output option 1 :
```
{
	"packageRules": [
		{
			"matchPackageNames": [
				"@floating-ui/react-dom"
			],
			"allowedVersions": "<0.6.3"
		},
		{
			"updateTypes": [
				"major"
			],
			"labels": [
				"UPDATE-MAJOR"
			]
		}
	],
	"ignoreDeps": [
		"next"
	]
}
```
Output option 2 :
Or should the expected print just be the `repo-a` central config which we are not differentiating

```
{
	"packageRules": [
		{
			"matchPackageNames": [
				"@floating-ui/react-dom"
			],
			"allowedVersions": "<0.6.3"
		},
		{
			"updateTypes": [
				"major"
			],
			"labels": [
				"UPDATE-MAJOR"
			]
		}
	]
}
```

**Comment by rarkins** [comment]
All internal presets should be retained in their `extends` arrays, e.g.

Repo b:

```
{
        "extends": ["config:base"],
	"packageRules": [
		{
			"matchPackageNames": [
				"@floating-ui/react-dom"
			],
			"allowedVersions": "<0.6.3"
		},
		{
			"updateTypes": [
				"major"
			],
			"labels": [
				"UPDATE-MAJOR"
			]
		}
	],
	"ignoreDeps": [
		"next"
	]
}
```

**Comment by rarkins** [comment]
Logging itself should be done in `workers/repository/init/merge.ts`, i.e. same place we log raw and migrated config.

**Comment by nabeelsaabna** [comment]
calling the `presets.resolveConfigPresets()` twice might result in doubling the HTTP calls
Should we implement some caching to prevent this?

**Comment by rarkins** [comment]
Yes if that's the case and we are not caching currently then we should add some cached resolution to the in-memory repo cache (separate PR preferred).

**Comment by hasanwhitesource** [comment]
The main concern was performance related to http calls, now its more clear. I will now update the code accordingly.

**Comment by rarkins** [comment]
Use the [memory cache](https://github.com/renovatebot/renovate/blob/main/lib/util/cache/memory/index.ts) so that it's cleared for each repo.
